### PR TITLE
fix: Omit type antecedents for unassigned variables

### LINKED ISSUE #1958: Empty type of let variable assumed before binding
The following program verifies, but should not:

``` dafny
type R = r: BaseType | r.i() witness *

function F(v: nat): R {
  var o: R := Ctor(v);
  o
}
```

The following similar code, but without the let expression, does produce an error (which is good):

``` dafny
function G(v: nat): R {
  Ctor(v) // error: This does not produce an R
}
```

A quick look at the generated Boogie shows that `var o: R ...` generates an assumption about `o`'s type before the RHS has been evaluated and checked. That assumption should not be there (at least for a possibly empty type).

### PR DESCRIPTION
**Pull request by RustanLeino**
In let expressions and named function results, the bound variable was previously introduced with an assumption that the variable satisfied its type. That is not sound, since the type may be empty. This PR removes those premature type assumptions.

Fixes #1958

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-1958.dfy [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    else if s[2] == x then
+      Some(2)
+    else
+      None
+  }
+
+  function Search4<T(==)>(s: seq<T>, x: T): (o: Option<int>)
+    ensures
+      o.Some? && o.value < |s| ==>
+      s[o.value] == x // error: ".value" may be negative
```
Comment (MikaelMayer) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-1958.dfy:167]: How does this function relates to the soundness issue being solved? Is that the baseline without subset types?

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-1958.dfy.expect [hunk 2]
```diff
@@ -0,0 +1,19 @@
+git-issue-1958.dfy(13,16): Error: value does not satisfy the subset constraints of 'R'
+git-issue-1958.dfy(18,16): Error: value does not satisfy the subset constraints of 'R'
+git-issue-1958.dfy(32,4): Error: value does not satisfy the subset constraints of 'R'
+git-issue-1958.dfy(36,4): Error: value does not satisfy the subset constraints of 'R'
+git-issue-1958.dfy(56,16): Error: function precondition could not be proved
```
Comment (MikaelMayer) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-1958.dfy.expect:5]: That one, along with the one line 104, is not triggered on my machine and that's a soundness issue. So you fixed it !

**Review by MikaelMayer — APPROVED** [review]
Looks like this PR does the job! Well done.

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-1958.dfy (hunk 1 above)
Comment (RustanLeino) [reply Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-1958.dfy:167]: Yes, that's the baseline -- just to make sure there's nothing else that causes the other tests to pass.

**Pull request merged**